### PR TITLE
New multi-colored worklist cuda analysis

### PR DESCRIPTION
**Pull request by thorulf4**

### CODE REVIEW

File: src/cuda/bit_vector_converter.h [hunk 1]
```diff
@@ -15,4 +16,21 @@ void set_bit_cuda_state(std::vector<NodeType>& nodes, std::map<std::string, int>
             }
         }                
     }   
+}
+
+template<typename NodeType>
+void set_bit_cuda_multi_state(DynamicArray<NodeType>& nodes, std::map<std::string, int>& variables, int source_count, 
+                   std::vector<StatefulNode<SourcedTaintState>>& cfg_nodes){
```
Comment (NicklasHugoy) [inline src/cuda/bit_vector_converter.h:23]: Does it really make sense to make this templated?

**Review by ZcortezZ — CHANGES_REQUESTED** [review]

File: src/cfg/transformations/taint_locator.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
+#include "taint_locator.h"
+#include <cfg/Expression.h>
+#include <memory>
+
+bool TaintSourceLocator::is_taintsource(Node& node){
+    node.accept(*this);
+    return last_is_taintsource;
+}
```
[the remark this answers is not in the document]
Comment (NicklasHugoy) [reply src/cfg/transformations/taint_locator.cpp:8]: or keep the method for easier to read code but change it to `return node.accept(*this)`

File: src/cuda/multi_cuda/analysis.cu [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            
+
+            if(++collision_count >= COLLISIONS_BEFORE_SWITCH){
+                current_work_column = (current_work_column + 1) % work_column_count;
+                work_column = work_columns[current_work_column];
+            }else{
+                hash++;
+            }
+        }
+        work_column[hash % THREAD_COUNT] = succ_index;
```
Comment (NicklasHugoy) [inline src/cuda/multi_cuda/analysis.cu:41]: Opportunity to reuse code from cuda-worklist?

File: src/main.cpp [hunk 4]
```diff
@@ -236,10 +264,11 @@ int main(int argc, char *argv[]){
             std::cout << "Invalid command\n";
             std::cout << " --cpu -c for use on cpu\n";
             std::cout << " with cpu flag:\n";
-            std::cout << "  --multi -m for multi taint analysis\n";
+            std::cout << " --multi -m for multi taint analysis\n";
             std::cout << " --gpu -g for use on gpu\n";
             std::cout << " --cuda -cu for bit-cuda implementation\n";
             std::cout << " --cuda-worklist -cw for bit-cuda implementation using worklist\n";
+            std::cout << " --multi-cuda -mc for multi colored taint analysis using cuda and worklist\n";
```
Comment (NicklasHugoy) [inline src/main.cpp:271]: Update README with the new multi-cuda flag as well

**Review by NicklasHugoy — CHANGES_REQUESTED** [review]

File: src/cuda/multi_cuda/analysis.cu (hunk 3 above)
Comment (thorulf4) [reply src/cuda/multi_cuda/analysis.cu:41]: I was planning on getting it working first then moving into refactoring and finding some way to reuse the cuda code.

**Review by ZcortezZ — DISMISSED** [review]

**Review by NicklasHugoy — DISMISSED** [review]

**Review by NicklasHugoy — DISMISSED** [review]

**Review by ZcortezZ — APPROVED** [review]

**Review by ZcortezZ — APPROVED** [review]

**Pull request merged**